### PR TITLE
Coverage parameter scenario

### PR DESCRIPTION
**Pull request by amuolo**
This PR is to introduce the possibility to import reinsurance coverage parameters for different scenarios.

### CODE REVIEW

File: ifrs17/Utils/EqualityComparers.ipynb [hunk 1]
```diff
@@ -140,6 +140,35 @@
...
+                "\n",
+                "\n    public bool Equals(DataNodeParameter x, DataNodeParameter y)",
+                "\n    {",
+                "\n        if (x == null && y == null) return true; ",
+                "\n        if (x == null || y == null) return false; ",
+                "\n        if (!(x.Year == y.Year && x.Month == y.Month && x.DataNode == y.DataNode && x.Scenario == y.Scenario)) return false; ",
+                "\n        if (x is SingleDataNodeParameter && y is SingleDataNodeParameter &&",
+                "\n            ((SingleDataNodeParameter)x).PremiumAllocation == ((SingleDataNodeParameter)y).PremiumAllocation) return true; ",
+                "\n        if (x is InterDataNodeParameter && y is InterDataNodeParameter &&",
+                "\n            ((InterDataNodeParameter)x).LinkedDataNode == ((InterDataNodeParameter)y).LinkedDataNode && ",
```
Comment (dcolleoni) [inline ifrs17/Utils/EqualityComparers.ipynb:158]: DataNode and LinkedDataNode might be inverted but the comparer should return true

File: ifrs17/Import/Importers.ipynb [hunk 2]
```diff
@@ -503,18 +503,33 @@
...
                 "\n    switch(format)",
                 "\n    {",
-                "\n        case ImportFormats.YieldCurve : ",
+                "\n        case ImportFormats.Cashflow:",
+                "\n        case ImportFormats.Actual: ",
+                "\n        case ImportFormats.Opening:",
+                "\n        case ImportFormats.SimpleValue:",
+                "\n        {",
+                "\n            if(args.Scenario != null) return args.RepeatOnce().ToArray();",
+                "\n            var secondaryArgs = await dataSource.Query<PartitionByReportingNodeAndPeriod>() ",
```
[the remark this answers is not in the document]
Comment (dtrzesniak) [reply ifrs17/Import/Importers.ipynb:517]: Should it go the parsing storage and cached?

[the remark this answers is not in the document]
Comment (amuolo) [reply ifrs17/Import/Importers.ipynb:517]: Yes, I agree. If we used the parsing storage solely to query the database only once (per type), then we would recycle the data without querying multiple times. This is doable but it would require quite a few changes in the importers: both in the parsing storage to ensure that all needed types are queried, and also in these utility helper methods to be able to pass the parsing storage as parameter.

[the remark this answers is not in the document]
Comment (amuolo) [reply ifrs17/Import/Importers.ipynb:517]: I suggest to implement these changes in a separate PR. Can I?

File: ifrs17/Import/Importers.ipynb [hunk 3]
```diff
@@ -1180,13 +1174,52 @@
...
+                "\n        if(ApplicationMessage.HasErrors()) return Activity.Finish().Merge(log);",
+                "\n        ",
+                "\n        // Avoid starting the computation if no best estimate cash flow and actuals has ever been imported ",
+                "\n        if(args.Scenario == null) {",
+                "\n            if(!(await options.TargetDataSource.Query<RawVariable>().Where(x => x.Partition == targetPartition).Take(1).ToArrayAsync()).Any() &&",
+                "\n               !(await options.TargetDataSource.Query<IfrsVariable>().Where(x => x.Partition == targetPartition).Take(1).ToArrayAsync()).Any()) continue;",
+                "\n        }",
+                "\n",
+                "\n        // Remove data nodes which are unaffected by the updated yield curves",
+                "\n        var nominals = await options.TargetDataSource.Query<RawVariable>().Where(x => targetDataNodes.Contains(x.DataNode) && (x.Partition == targetPartition || x.Partition == defaultPartition)).ToArrayAsync();",
```
[the remark this answers is not in the document]
Comment (amuolo) [reply ifrs17/Import/Importers.ipynb:1213]: That's right! Done!

File: ifrs17/Import/Importers.ipynb (hunk 2 above)
[the remark this answers is not in the document]
Comment (amuolo) [reply ifrs17/Import/Importers.ipynb:517]: I created this ticket to this aim
https://github.com/Systemorph/IFRS17CalculationEngine/issues/237

**Review by dtrzesniak — APPROVED** [review]

File: ifrs17/Import/Importers.ipynb [hunk 4]
```diff
@@ -503,18 +503,33 @@
...
-                "\n        case ImportFormats.YieldCurve : ",
+                "\n        case ImportFormats.Cashflow:",
+                "\n        case ImportFormats.Actual: ",
+                "\n        case ImportFormats.Opening:",
+                "\n        case ImportFormats.SimpleValue:",
+                "\n        {",
+                "\n            if(args.Scenario != null) return args.RepeatOnce().ToArray();",
+                "\n            var secondaryArgs = await dataSource.Query<PartitionByReportingNodeAndPeriod>() ",
+                "\n                .Where(x => x.ReportingNode == args.ReportingNode && x.Year == args.Year && x.Month == args.Month && x.Scenario != null)",
+                "\n                .Select(x => new ImportArgs(x.ReportingNode, x.Year, x.Month, default(Periodicity), x.Scenario, ImportFormats.Cashflow)).ToArrayAsync();",
```
Comment (dcolleoni) [inline ifrs17/Import/Importers.ipynb:519]: ImportFormat is set to ImportFormats.Cashflow in every case? also when ImportFormats is Actual and Opening?

File: ifrs17/Import/Importers.ipynb [hunk 5]
```diff
@@ -503,18 +503,33 @@
...
+                "public async Task<ImportArgs[]> GetAllArgsAsync(ImportArgs args, IDataSource dataSource, string format)",
                 "\n{",
                 "\n    ImportArgs[] allArgs;",
                 "\n    switch(format)",
                 "\n    {",
-                "\n        case ImportFormats.YieldCurve : ",
+                "\n        case ImportFormats.Cashflow:",
+                "\n        case ImportFormats.Actual: ",
+                "\n        case ImportFormats.Opening:",
+                "\n        case ImportFormats.SimpleValue:",
```
Comment (dcolleoni) [inline ifrs17/Import/Importers.ipynb:514]: why SimpleValue is here? this format does not compute anything.

Comment (dtrzesniak) [reply ifrs17/Import/Importers.ipynb:514]: Good point.

Comment (amuolo) [reply ifrs17/Import/Importers.ipynb:514]: yes, agree, we can remove simple value.

File: ifrs17/Import/Importers.ipynb (hunk 4 above)
Comment (amuolo) [reply ifrs17/Import/Importers.ipynb:519]: I believe this is correct because the secondary args are all those scenario args that come into play when importing something Best Estimate. In this implementation, scenario calculations have import format cashflow.

Comment (amuolo) [reply ifrs17/Import/Importers.ipynb:519]: This line has not been changed by this PR, that is, same behavior was there prior to this PR.

File: ifrs17/Import/Importers.ipynb [hunk 6]
```diff
@@ -1180,13 +1173,52 @@
...
+                "\n",
+                "\n    if(log.Errors.Any()) return Activity.Finish().Merge(log); ",
+                "\n    var toCommitParameters  = (await workspace.Query<DataNodeParameter>().ToArrayAsync()).Except(committedParameters, ParametersComparer.Instance());",
+                "\n    if (!toCommitParameters.Any()) {",
+                "\n        ApplicationMessage.Log(Warning.VariablesAlreadyImported);   ",
+                "\n        return Activity.Finish().Merge(log);",
+                "\n    }",
+                "\n",
+                "\n    var allArgs = await GetAllArgsAsync(primaryArgs, options.TargetDataSource, ImportFormats.DataNodeParameter);",
+                "\n    var targetDataNodes = toCommitParameters.Select(x => x.DataNode).ToHashSet();",
```
Comment (dcolleoni) [inline ifrs17/Import/Importers.ipynb:1194]: For InterDataNodeParameter you also have a linkedDataNode. Why is this not a targetDataNode?

File: ifrs17/Import/Importers.ipynb (hunk 4 above)
Comment (dcolleoni) [reply ifrs17/Import/Importers.ipynb:518]: Having FormatCashflow computes from RawVar and it does not compute Actuals such as Advance and Overdue. 
If you re-import an Actual BE and you have already imported an Actual scenario you need to recompute the scenario performing calculation of Advance and Overdue. Now I think you would re-compute them.

Comment (amuolo) [reply ifrs17/Import/Importers.ipynb:518]: Do you suggest to compute from Actual in case of a (re-)import of an Actual BE?

File: ifrs17/Import/Importers.ipynb (hunk 6 above)
Comment (amuolo) [reply ifrs17/Import/Importers.ipynb:1194]: Because the linked data nodes are retrieved from the secondary scope of the import storage employing the relaxed logic. I think this might be correct, right?

File: ifrs17/Import/Importers.ipynb (hunk 4 above)
Comment (amuolo) [reply ifrs17/Import/Importers.ipynb:518]: ok, I am convinced and I agree with you: the import format in the case of BE actuals need to be ImportFormats.Actual. I have applied the change to achieve this behavior in the lines you picked up.

File: ifrs17/Import/Importers.ipynb (hunk 6 above)
Comment (amuolo) [reply ifrs17/Import/Importers.ipynb:1194]: You are super right! If I do not consider also the LinkedDataNode and upload them too to the workspace, the import storage wouldn't have a chance to retrieve the corresponding IfrsVariables or RawVariables. Now I fixed this. Thanks!

File: ifrs17/Import/Importers.ipynb [hunk 7]
```diff
@@ -503,18 +503,33 @@
...
                 "\n    ImportArgs[] allArgs;",
                 "\n    switch(format)",
                 "\n    {",
-                "\n        case ImportFormats.YieldCurve : ",
+                "\n        case ImportFormats.Cashflow:",
+                "\n        case ImportFormats.Actual: ",
+                "\n        case ImportFormats.Opening:",
+                "\n        {",
+                "\n            if(args.Scenario != null) return args.RepeatOnce().ToArray();",
+                "\n            var importFormat = (format == ImportFormats.Actual)? ImportFormats.Actual : ImportFormats.Cashflow;",
```
Comment (dcolleoni) [inline ifrs17/Import/Importers.ipynb:516]: Now I think Opening also imports Advance and Overdue, and you would then need to leave format = opening else with Cashflow you would have the same problem we discussed for actuals.

**Comment by dcolleoni** [comment]
Why do we have so many changes in BMs? is it because of the format we fixed?

File: ifrs17/Import/Importers.ipynb (hunk 7 above)
Comment (amuolo) [reply ifrs17/Import/Importers.ipynb:516]: ok done

**Comment by amuolo** [comment]
> Why do we have so many changes in BMs? is it because of the format we fixed?

Not only, it is also because of line 570 of the Importers notebook, that is, variables scenarios that are equal to BE are not saved anymore. Before we didn't have this additional filter and we duplicated the values.

File: ifrs17/Import/Importers.ipynb [hunk 8]
```diff
@@ -503,18 +503,32 @@
...
+                "\n                         .Distinct();",
                 "\n                var targetPartitions = await dataSource.Query<PartitionByReportingNodeAndPeriod>()",
-                "\n                    .Where(x => x.Year == args.Year && x.Month == args.Month && !scenariosWithYieldCurves.Contains(x.Scenario)).OrderBy(x => x.Scenario).ToArrayAsync();",
+                "\n                    .Where(x => x.Year == args.Year && x.Month == args.Month && !scenarios.Contains(x.Scenario)).OrderBy(x => x.Scenario).ToArrayAsync();",
                 "\n                var targetScenarios = targetPartitions.Where(x => x.Scenario != null).Select(x => x.Scenario);",
                 "\n                if(targetScenarios.Any()) ApplicationMessage.Log(Warning.ScenarioReCalculations, String.Join(\", \", targetScenarios));",
                 "\n                allArgs = targetPartitions.Select(x => new ImportArgs(x.ReportingNode, x.Year, x.Month, default(Periodicity), x.Scenario, ImportFormats.Cashflow)).ToArray();",
```
Comment (dcolleoni) [inline ifrs17/Import/Importers.ipynb:534]: the format is set to Cashflow here. While for YC I see as necessary, for a generic ParameterUpdated it might not be necessary. Given the parameters we have now it would be enough format Actual. This would speed up import. On the other side the code would get more complex having to decide which format is the best given the scenario. Also, we might add new parameters in the future making the logic even more complex. 
In summary, I think it is a good decision to keep Cashflow format here.

File: ifrs17/Import/Importers.ipynb [hunk 9]
```diff
@@ -1180,13 +1174,53 @@
...
+                "\n        if(ApplicationMessage.HasErrors()) return Activity.Finish().Merge(log);",
+                "\n        ",
+                "\n        // Avoid starting the computation if no best estimate cash flow and actuals has ever been imported ",
+                "\n        if(args.Scenario == null) {",
+                "\n            if(!(await options.TargetDataSource.Query<RawVariable>().Where(x => x.Partition == targetPartition).Take(1).ToArrayAsync()).Any() &&",
+                "\n               !(await options.TargetDataSource.Query<IfrsVariable>().Where(x => x.Partition == targetPartition).Take(1).ToArrayAsync()).Any()) continue;",
+                "\n        }",
+                "\n",
+                "\n        // Only nominals corresponding to the target data nodes are added to the workspace",
+                "\n        var nominals = await options.TargetDataSource.Query<RawVariable>().Where(x => targetDataNodes.Contains(x.DataNode) && (x.Partition == targetPartition || x.Partition == defaultPartition)).ToArrayAsync();",
```
Comment (dcolleoni) [inline ifrs17/Import/Importers.ipynb:1214]: don't you need also previous period partition?
This solution seems to me a bit complex... and it is too easy to miss some pieces... also hard to maintain...

File: ifrs17/Import/Importers.ipynb (hunk 8 above)
Comment (amuolo) [reply ifrs17/Import/Importers.ipynb:535]: Yes, thanks! I also think it is quite a general rule that would remain valid for more parameters.

File: ifrs17/Import/Importers.ipynb (hunk 9 above)
Comment (amuolo) [reply ifrs17/Import/Importers.ipynb]: previous partition added! We can improve this logic with the next PR.

**Review by dcolleoni — APPROVED** [review]
